### PR TITLE
add MockBigQuery and integration tests

### PR DESCRIPTION
**Pull request by nevillelyh**
Proof of concept but seems to work well. Wanna get more feedbacks on API usability, edge cases etc.

### CODE REVIEW

File: scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+/**
+ * BigQuery integration test with mock data and equality check for query result.
+ */
+class BigQueryMockIT extends FlatSpec with Matchers with BigQueryITUtils {
+
+  import BigQueryIT._
+
+  private def shakespeare(w: String, wc: Long, c: String, cd: Long) =
+    TableRow("word" -> w, "word_count" -> wc, "corpus" -> c, "corpus_date" -> cd)
+
```
Comment (andrewsmartin) [inline scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala]: Would it be worth adding some helper so that users don't have to define their own functions / create table rows manually? Realize it's not much overhead really, but even just so that people don't have to work with TableRow directly. Maybe something like this? Would have to define this for up to 23 type params but could easily be generated similar to MultiJoin, and it's slightly "safer" in that users won't be able to pass in TableRows with different arity.

```
mockTable(tableRef).withData(
  Data("word", "word_count", "corpus", "corpus_date")
    .add("i", 10, "kinglear", 1600)
    .add("thou", 20, "kinglear", 1600)
  )
)
```

Comment (andrewsmartin) [inline scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala]: Then again all of this is just solved by using typed data anyway.

File: scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  it should "fail duplicate mock data" in {
+    val it = new BigQueryITUtils {}
+    val mt = it.mockTable(tableRef)
+    mt.withData(Nil)
+    the[IllegalArgumentException] thrownBy {
+      mt.withData(Nil)
+    } should have message s"requirement failed: Table $tableRef already populated with mock data"
+  }
+
```
Comment (andrewsmartin) [inline scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala]: If we register a mock table but don't supply any data, what's the expected behavior? Treat it as empty table and still allow querying? Should we include a test for this?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    val t = bq.getTable(original)
+    val temp = bq.temporaryTable(t.getLocation)
+    mapping += (original -> temp)
+    new MockTable(bq, t.getSchema, original, temp)
+  }
+
+  /**
+   * Perform a live query against BigQuery service, substituting mocked tables with test data.
+   */
+  def query(sqlQuery: String, flattenResults: Boolean = false): Seq[TableRow] = {
```
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: Just a minor thing (and maybe it's just me) but would `runQuery()` or something be a better name? For me just `query()` makes it sound like a factory function to create some kind of query object instead of actually running one.

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    val isLegacy = bq.isLegacySql(sqlQuery, flattenResults)
+    val mockQuery = mapping.foldLeft(sqlQuery){ case (q, (src, dst)) =>
+      q.replace(toTableSpec(src, isLegacy), toTableSpec(dst, isLegacy))
```
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: What do we want to happen if the query hits tables that haven't been mocked? Should we warn if this is the case?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    bq.getQueryRows(mockQuery, flattenResults).toList
+  }
+
+  /**
+   * Perform a live query against BigQuery service, substituting mocked tables with test data.
+   */
+  def typedQuery[T <: HasAnnotation : ClassTag : TypeTag]
```
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: Same as above, is `runTypedQuery` a better name?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+   * Populate the table with mock data.
+   */
+  def withTypedData[T <: HasAnnotation : ClassTag : TypeTag](rows: Seq[T]): Unit = {
+    val bqt = BigQueryType[T]
+    withData(rows.map(bqt.toTableRow))
+  }
+
+  /**
+   * Populate the table with sample data from the original table.
+   * Note that rows are taken from the beginning of the table and not truly random.
```
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: I can imagine this being a problem with a lot of tables - could we use reservoir sampling? Or should we wait until someone asks for it :)

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala (hunk 3 above)
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: I guess it is also a verb

File: scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala (hunk 2 above)
Comment (nevillelyh) [inline scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala]: The query will fail but we can't ensure that the user actually calls `with*`.

File: scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala (hunk 1 above)
Comment (nevillelyh) [inline scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala]: Yeah not a problem really. More API = more confusion.

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala (hunk 3 above)
Comment (nevillelyh) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: I'm leaning towards a noun since it's used in contexts like `query("...") should contain`.
I'll rename it to `queryResult`?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala (hunk 4 above)
Comment (nevillelyh) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: Maybe but we'll have to extract tables using regex which is hacky. I think it's ok to hit real tables (e.g. small lookup ones) but don't wanna go crazy with query parsing.

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala (hunk 6 above)
Comment (nevillelyh) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala]: Let's wait, think of this as a MVP :)
